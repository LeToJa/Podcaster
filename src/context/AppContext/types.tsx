export interface AppContextTypes {
    loading: boolean
    toggleLoading: (value: boolean) => void
}