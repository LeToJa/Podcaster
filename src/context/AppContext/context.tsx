import { createContext, useState } from 'react'

import { ReactElementWithChildrenTypes } from '../../types'
import { AppContextTypes } from './types'

export const AppContext = createContext<AppContextTypes | undefined>(undefined)

export const AppProvider = ({ children }: ReactElementWithChildrenTypes) => {
    const [loading, setLoading] = useState<boolean>(true)

    const toggleLoading = (value: boolean) => setLoading(value)

    return <>
        <AppContext.Provider value={{ loading, toggleLoading }}>
            {children}
        </AppContext.Provider>
    </>
}