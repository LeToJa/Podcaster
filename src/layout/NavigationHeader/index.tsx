import { Link } from 'react-router-dom'

import { useApp } from '../../context/AppContext'

const NavigationHeader = () => {
    const { loading } = useApp()

    return <>
        <nav className='flex justify-between mb-10' data-testid='app-navbar'>
            <Link to='/' className='font-bold text-xl text-sky-600' data-testid='navbar-title'>Podcaster</Link>
            {!!loading && <figure className='animate-ping w-2 h-2 rounded-full bg-sky-600' data-testid='navbar-indicator'></figure>}
        </nav>
    </>
}

export default NavigationHeader