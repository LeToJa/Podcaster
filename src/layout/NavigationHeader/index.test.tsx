import { render } from '@testing-library/react'
import { describe, it, expect, vi, Mock } from 'vitest'
import { BrowserRouter } from 'react-router-dom'

import NavigationHeader from '.'
import { useApp } from '../../context/AppContext'

vi.mock('../../context/AppContext', () => ({
    useApp: vi.fn(),
}))

describe('NavigationHeader', () => {
    it('renders correctly and matches snapshot', () => {
        (useApp as Mock).mockReturnValue({ loading: false })

        const { container } = render(
            <BrowserRouter>
                <NavigationHeader />
            </BrowserRouter>
        )

        expect(container).toMatchSnapshot()
    })

    it('renders the Podcaster link', () => {
        (useApp as Mock).mockReturnValue({ loading: false })

        const { getByText } = render(
            <BrowserRouter>
                <NavigationHeader />
            </BrowserRouter>
        )

        const linkElement = getByText('Podcaster')

        expect(linkElement).toBeTruthy()
        expect(linkElement).toHaveProperty('href', 'http://localhost:3000/')
    })

    it('does not render the loading indicator when loading is false', () => {
        (useApp as Mock).mockReturnValue({ loading: false })

        const { queryByRole } = render(
            <BrowserRouter>
                <NavigationHeader />
            </BrowserRouter>
        )

        expect(queryByRole('figure')).toBeFalsy()
    })

    it('renders the loading indicator when loading is true', () => {
        (useApp as Mock).mockReturnValue({ loading: true })

        const { queryByRole } = render(
            <BrowserRouter>
                <NavigationHeader />
            </BrowserRouter>
        )

        expect(queryByRole('figure')).toBeTruthy()
    })
})
