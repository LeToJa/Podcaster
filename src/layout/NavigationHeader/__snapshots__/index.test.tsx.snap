// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavigationHeader > renders correctly and matches snapshot 1`] = `
<div>
  <nav
    class="flex justify-between mb-10"
    data-testid="app-navbar"
  >
    <a
      class="font-bold text-xl text-sky-600"
      data-testid="navbar-title"
      href="/"
    >
      Podcaster
    </a>
  </nav>
</div>
`;
